### ISSUE TITLE
What Triggers 'Significant' and 'Meaningless' Changes?

### ISSUE THREAD

**Issue by PhillipFraThailand** [callstack/reassure-examples#21]
Hello everyone 👋 Huge shoutout to CallStack for creating reassure!

I am looking into integrating reassure into my teams pipelines, and creating a white label that automatically generates performance and unit tests for all new components created for my company to use in future projects for our clients.

But in order to do that, i would like to know more about what would cause a change to be considered significant or meaningless.

I did some quick testing, and found that i could increase rendering count and time of components a lot, before the changes would be considered significant.

I have been looking for information on this online and listened to the podcast but were not able to find any details on this.

My goal is to be able to give examples to my team on what code examples and scenarios that would cause the changes to be considered significant.

For example, starting out with a simple component, adding useEffects, states and calculations and investigating when it's okay to ignore the warnings in the pipeline and when not to.

Can anyone provide me some information about this, or point me in the direction of where i might learn more about this?

**Comment by thymikee** [comment]
Hey! The code is here: https://github.com/callstack/reassure/blob/99afdf9848ca48eeab27091ce3cb5e11e57b20b6/packages/reassure-compare/src/compare.ts#L18-L21

I don't remember where these came from, probably from V8 codebase. @mdjastrzebski would know better here, and I hope he'll be able to answer next week :)

**Comment by PhillipFraThailand** [comment]
Hey @thymikee, thanks for the fast reply!

I've had the day to look into the inner workings of Reassure, to air it here to hopefully get corrected where i am wrong, but also to help you provide me with a better answer and hopefully also benefit other people who arrive here.

For rendition time ms to be considered significant, the changes has to pass two conditions: 
- [The mean rendition time must be higher than a hardcoded minimum threshold of 4ms](https://github.com/callstack/reassure/blob/99afdf9848ca48eeab27091ce3cb5e11e57b20b6/packages/reassure-compare/src/compare.ts#L29C47-L29C47).
- The test values must be statistically relevant, meaning that, in this case, the tests results isn't due to an unstable environment. To do this, Reassure calculates Z-scores for all tests, then uses a standard Z-score to P-value, lookup table, to check if the test result is below the [chosen acceptable threshold](https://github.com/callstack/reassure/blob/99afdf9848ca48eeab27091ce3cb5e11e57b20b6/packages/reassure-compare/src/compare.ts#L21).

I have been able to verify this in my own tests by just adding components to my view until the rendition time came above 4ms.

However, for the rendering count, it seems that it could never trigger a significant change by itself? I was not able to increase it to a point where my changes would be considered significant, before i also increased the rendition time to above 4ms.
[I found this part of the code](https://github.com/callstack/reassure/blob/99afdf9848ca48eeab27091ce3cb5e11e57b20b6/packages/reassure-compare/src/compare.ts#L161) that uses the [rendition count threshold](https://github.com/callstack/reassure/blob/99afdf9848ca48eeab27091ce3cb5e11e57b20b6/packages/reassure-compare/src/compare.ts#L35) but by looking at the code, and trying to provoke it in my tests, am i correct to say that render count could never be considered a significant change?

If so, im curious as to to what the reasoning behind that might be? As my understanding is that unnecessary rerenders  could impact performance. Could it be because, the rendering time already is descriptive enough of the performance? But wouldn't it be nice to be able to warn about this, as it potentially could lead to rerendering of the components children later?

**Comment by mdjastrzebski** [comment]
@PhillipFraThailand your research seems right. Let me rephrase it and add some context to it, so that we are on the same page:

**Significant / Meaningless Render Duration**
* the basis for significant / meaningless categorization is the statistical hypothesis tests (z score, etc) based on mean durations from baseline and current, as well as std dev & number of runs. This is based on V8 perf testing method. The goal her is to analyse not only the mean values but also their dispersion and count to be able to establish if these two sets of results are statistically different given used certainly parameter.
* the 4 ms threshold is additional check added, as the grain on measurement is 1 ms. This seams that for very short render durations, is easy to between e.g. 1 and 2 ms, which would easily trigger the first check. Therefore we have this 4ms threshold which will limit such small value cases. For best stability, your renders should take at least ~100ms each.

**Render Count**
* render count is a different type of check, not related to significant check above. It is intended to help users catch cases when they inadvertently introduced additional render to their components, e.g. by breaking the memoization, etc. Render counts are more stable measure then render duration, as typically your component has fixed number of renders under given scenario. 

@PhillipFraThailand hope that answers your questions. If I left some piece unclear please add more specific questions.

**Comment by PhillipFraThailand** [comment]
Hey @mdjastrzebski thanks for the clarification and sorry for my late response!

Just one last question, will it be possible, sometime in the future, to pass settings for what should be considered significant, or potentially adding Render count also? 

We might add that into our pipelines ourselves for now,  i will share it here when we get further with this :) 

Thanks!

**Comment by mdjastrzebski** [comment]
@PhillipFraThailand customizing the "siginficant/meaningless" condition could be an option for the future, but tbh I do not have idea how to make it a usable API. Perhaps you can submit some proposal for that.

That being said, Reassure beside creating Markdown file also creates a JSON file containing all the relevant data for each test: invidual runs, means, std dev, for both baseline and current. You could use that data to e.g. reprocess the file and change the classification of result. 

Basically each tests get classified as exactly one of `significant`, `meaningless`, `added` (no baseline entry) or `removed` (no current entry). The entries in `countChanged` will contain duplicates of the above.

**Comment by mdjastrzebski** [comment]
Closing as stale.
